1기 2회차 문상희 20220805 모의고사 풀이 완료하고 소감작성했습니다.

MoonSanghee opened the pull request.

hphktutor1 commented on 2회차/문상희/0_민석이의과제체크하기.py:
```diff
@@ -1,3 +1,19 @@
+    print(f'#{test}', end = ' ')
+    for i in do_not:
+        print(i, end = ' ')
+    print()
```
잘 푸셨습니다!

12~14번째 줄의 시간 복잡도에 대해서 살펴보면,
people이 N, done_list도 최대 N이기 때문에 O(N**2)이 됩니다.

이를 O(N)으로 바꾸는 방법이 있습니다!
미리 새로운 리스트나 딕셔너리에 수업을 들은 학생을 체크한 후에,
1번부터 N번까지 학생을 순회하면서 수업을 듣지 않은 학생을 프린트 하는 것입니다.

hphktutor1 replied:
잘 푸셨습니다! 
12~14번째 줄의 시간 복잡도에 대해서 살펴보면, people이 N, done_list도 최대 N이기 때문에 O(N**2)이 됩니다. 

이를 O(N)으로 바꾸는 방법이 있습니다!
미리 새로운 리스트나 딕셔너리에 수업을 들은 학생을 체크한 후에, 1번부터 N번까지 학생을 순회하면서 수업을 듣지 않은 학생을 프린트 하는 것입니다.

hphktutor1 commented on 2회차/문상희/1_조교의성적매기기.py:
```diff
@@ -1,3 +1,27 @@
+    re_point = sorted(point_list, reverse = True)
+    a = re_point.index(finding)
+    # 성적을 모은 리스트를 내림차순으로 정렬하여 찾은 점수의 인덱스 값으로 다시 만들어준 glade2 리스트에서 성적을 찾습니다.
+    print(f'#{test} {glade2[a]}')
```
잘 푸셨습니다!👍

hphktutor1 commented on 2회차/문상희/2_암호생성기.py:
```diff
@@ -1,3 +1,24 @@
+    print(f'#{test}', end = ' ')
+    for i in numbers:
+        print(i, end = ' ')
+    print()
```
잘 푸셨습니다!👍
리스트 슬라이싱을 활용해서 큐를 구현해 주셨군요!
`deque`를 활용한 풀이도 한번 해보시면 좋을 것 같습니다.

hphktutor1 commented on 2회차/문상희/3_파리퇴치.py:
```diff
@@ -1,3 +1,32 @@
+            # fly와 max를 비교하여 fly가 크면 max를 갱신하여 줍니다.
+
+    # 모든 반복이 끝나면 최댓값을 확인 할 수 있습니다.
+    print(f'#{test} {max}')
```
깔끔한 풀이입니다!
15번째 줄에서 max는 내장함수로 정의되어 있기 때문에, 변수명을 바꿔주면 더 좋겠네요!

hphktutor1 commented on 2회차/문상희/4_어디에단어가들어갈수있을까.py:
```diff
@@ -1,3 +1,55 @@
+    # 위와 동일한 방법으로 열을 순회하며 원하는 길이의 단어가 들어갈 수 있는지 확인해주고 값을 추가해 줍니다.
+    # 행, 열에 각 몇개의 값을 원하는지가 아닌 총 갯수를 구하는 것이니 행에 구해준 값에 추가되는 값을 누적시켜 줍니다.
+            
+    print(f'#{test} {right}')
```
잘 푸셨네요!👍

27~31번째 줄의 경우에 `word = 0`을 하는 코드는 if일 때와 else일때 모두 들어가므로 else문을 없애고 `word = 0`을
밖으로 뺄 수도 있을 것 같네요!

hphktutor1 commented on 2회차/문상희/5_괄호짝짓기.py:
```diff
@@ -1,3 +1,46 @@
+    # 문자열를 다 돌았는데 리스트에 저장된 값이 있으면 닫히지 않은 괄호가 있으니 식이 성립할 수 없습니다.
+        else:    
+            print(f'#{test} 1')
+    # 이외의 경우는 식이 성립한기에 1을 출력해줍니다.
```
잘 푸셨습니다!👍

코드의 길이를 조금 줄여볼 수 있는 방법을 말씀드리면,
미리 `d = {')': '(', ']': '[', '}': '{', '>': '<'}` 와 같이 딕셔너리를 만들어 놓고
14~37번째 줄을 하나로 합칠 수 있습니다. 
```python
        else:
            if check and check[-1] == d.get(i):
                check.pop()
            else:
                print(f'#{test} 0')
                break
```

MoonSanghee commented:
새로운 리스트를 만들어 수업을 들은 학생을 체크하는것도 done_list를 순회하며 확인해 주어야하고 새로운 리스트를 확인하여 수업을
듣지 않은 학생을 찾는것도 n에 가까운 시간 복잡도를 가지는데 어떻게 o(n)의 복잡도로 줄어드는 건가요?

2022년 8월 9일 (화) 오후 3:13, hphktutor1 ***@***.***>님이 작성:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In 2회차/문상희/0_민석이의과제체크하기.py
> <https://github.com/kdt-hphk/01-PJT-04/pull/8#discussion_r940935688>:
>
> > +
> +test_case = int(input())
> +
> +for test in range(1, test_case + 1):
> +    people, done = map(int, input().split())
> +    done_list = list(map(int, input().split()))
> +    # 전체 수업 들은 사람은 range(1, people + 1)이고 과제를 제출한 사람이 몇명이고 누구인지 정보가 주어져있습니다.
> +    do_not = []
> +    for i in range(1, people + 1):
> +        if i not in done_list:
> +            do_not.append(i)
> +            # 그래서 빈 리스트를 만들고 전체를 순회하며 빈 리스트에 낸 사람에 포함되지 않은 사람들을 모아줍니다.
> +    print(f'#{test}', end = ' ')
> +    for i in do_not:
> +        print(i, end = ' ')
> +    print()
>
> 잘 푸셨습니다!
>
> 12~14번째 줄의 시간 복잡도에 대해서 살펴보면,
> people이 N, done_list도 최대 N이기 때문에 O(N**2)이 됩니다.
>
> 이를 O(N)으로 바꾸는 방법이 있습니다!
> 미리 새로운 리스트나 딕셔너리에 수업을 들은 학생을 체크한 후에,
> 1번부터 N번까지 학생을 순회하면서 수업을 듣지 않은 학생을 프린트 하는 것입니다.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kdt-hphk/01-PJT-04/pull/8#pullrequestreview-1066154797>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZ47AUTAF2JRY5W5CEPSSFLVYHZIDANCNFSM55U5CSNQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

MoonSanghee commented:
*미리 새로운 리스트나 딕셔너리에 수업을 들은 학생을 체크한 후에*, *1번부터 N번까지 학생을 순회하면서 수업을 듣지 않은 학생을
프린트 하는 것입니다.*

O(N)
                             O(N)

설명해 주신 부분에서 체크를 하는 과정이 O(N), 순회하며 아닌경우가 O(N)이라 O(N**2)의 복잡도를 가지게 되는게 아닌가요?



2022년 8월 10일 (수) 오후 1:01, hphktutor1 ***@***.***>님이 작성:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In 2회차/문상희/0_민석이의과제체크하기.py
> <https://github.com/kdt-hphk/01-PJT-04/pull/8#discussion_r941995306>:
>
> > +
> +test_case = int(input())
> +
> +for test in range(1, test_case + 1):
> +    people, done = map(int, input().split())
> +    done_list = list(map(int, input().split()))
> +    # 전체 수업 들은 사람은 range(1, people + 1)이고 과제를 제출한 사람이 몇명이고 누구인지 정보가 주어져있습니다.
> +    do_not = []
> +    for i in range(1, people + 1):
> +        if i not in done_list:
> +            do_not.append(i)
> +            # 그래서 빈 리스트를 만들고 전체를 순회하며 빈 리스트에 낸 사람에 포함되지 않은 사람들을 모아줍니다.
> +    print(f'#{test}', end = ' ')
> +    for i in do_not:
> +        print(i, end = ' ')
> +    print()
>
> 잘 푸셨습니다!
> 12~14번째 줄의 시간 복잡도에 대해서 살펴보면, people이 N, done_list도 최대 N이기 때문에 O(N**2)이 됩니다.
>
> 이를 O(N)으로 바꾸는 방법이 있습니다!
> 미리 새로운 리스트나 딕셔너리에 수업을 들은 학생을 체크한 후에, 1번부터 N번까지 학생을 순회하면서 수업을 듣지 않은 학생을 프린트
> 하는 것입니다.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/kdt-hphk/01-PJT-04/pull/8#pullrequestreview-1067632967>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AZ47AUTSROGFEQVTUW6BXFDVYMSSTANCNFSM55U5CSNQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
